### PR TITLE
Feature/intermediate-save

### PR DESCRIPTION
**Pull request by laurens-ritense**
…o carbon

### CODE REVIEW

File: projects/valtimo/config/assets/core/nl.json [hunk 1]
```diff
@@ -480,6 +480,13 @@
       "formBuilder": "Formulier editor",
       "jsonEditor": "JSON editor",
       "output": "Uitvoer"
+    },
+    "intermediateSave": {
+      "save": "Voortgang opslaan",
+      "clear": "Invoervelden wissen",
+      "clearConfirm": "Weet u zeker dat u uw invoervelden wilt wissen? Deze actie kan niet ongedaan worden.",
```
Comment (Rick-Ritense) [inline projects/valtimo/config/assets/core/nl.json]: I think the last sentence isn't completely correct

Deze actie kan niet ongedaan worden gemaakt.

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html [hunk 2]
```diff
@@ -13,16 +13,35 @@
...
-<valtimo-modal
+<cds-modal
   #taskDetailModal
-  elementId="taskDetailModal"
-  [title]="(page$ | async)?.title"
-  [subtitle]="(page$ | async)?.subtitle"
-  [templateBelowSubtitle]="assignUserToTask"
+  id="taskDetailModal"
 >
+  <cds-modal-header (closeSelect)="modal.open = false">
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html]: Add this logic in a method in the component

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html [hunk 3]
```diff
@@ -13,16 +13,42 @@
...
+            (click)="showConfirmationModal$.next(true)"
+          ></svg>
+        </cds-tooltip>
+      </div>
+    </div>
+  </cds-modal-header>
   <div
     body
+    class="cds--modal-content pb-1"
     *ngIf="{
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html:55]: You can add this whole *ngIf on the cds-modal and then you won't have to consume page$ with async in the header

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.scss [hunk 4]
```diff
@@ -19,3 +19,22 @@
     text-align: right;
   }
 }
+
+.cds-modal-header {
+  width: 100%;
+  display: flex;
+  flex-direction: row;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.scss]: flex-direction is row by default

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.scss [hunk 5]
```diff
@@ -19,3 +19,22 @@
...
+  justify-content: space-between;
+}
+
+.title-container {
+  width: 50%;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.scss:30]: Why is this necessary?

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html [hunk 6]
```diff
@@ -13,16 +13,42 @@
...
+        <cds-tooltip
+          [caret]="true"
+          [description]="'formManagement.intermediateSave.save' | translate"
+        >
+          <svg
+            class="cds--btn__icon save-icon-margin cursor-pointer"
+            cdsIcon="save"
+            size="16"
+            (click)="saveCurrentProgress()"
+          ></svg>
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html]: Turn this into a cdsButton with iconOnly set to true, then you won't need the save-icon-margin and cursor-pointer classes

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html [hunk 7]
```diff
@@ -13,16 +13,42 @@
...
+        <cds-tooltip
+          [caret]="true"
+          [description]="'formManagement.intermediateSave.clear' | translate"
+        >
+          <svg
+            class="cds--btn__icon cursor-pointer"
+            cdsIcon="trash-can"
+            size="16"
+            (click)="showConfirmationModal$.next(true)"
+          ></svg>
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html]: Same as above

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 8]
```diff
@@ -113,28 +115,32 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
...
+    //       tap(modalShowing => {
+    //         if (!modalShowing) {
+    //           if (this.formFlow) {
+    //             this.formFlow.saveData();
+    //           }
+    //         }
+    //       })
+    //     )
+    //     .subscribe()
+    // );
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]: Remove afterViewInit if the commented code is not needed

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 9]
```diff
@@ -158,12 +164,12 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
 
     //only load from formlink when process link failed for backwards compatibility
     if (!this.taskProcessLinkType$.getValue()) {
-      this.modal.show();
+      this.modal.open = true;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]: try this.modal = {...this.modal, open: true} 

Otherwise, change detection might fail in the future

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 10]
```diff
@@ -158,12 +164,12 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
...
     }
   }
 
   public gotoProcessLinkScreen(): void {
-    this.modal.hide();
+    this.modal.open = false;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]: Same as above

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 11]
```diff
@@ -205,7 +211,7 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
       this.toastr.success(
         `${task.name} ${this.translateService.instant('taskDetail.taskCompleted')}`
       );
-      this.modal.hide();
+      this.modal.open = false;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]: Same as above

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 12]
```diff
@@ -235,7 +242,7 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
               this.processLinkId$.next(res.processLinkId);
               this.formDefinition$.next(res.properties.formDefinition);
               this.formName$.next(res.properties.formName);
-              this.modal.show();
+              this.modal.open = true;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]: Same as above

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 13]
```diff
@@ -256,7 +263,7 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
   private setFormDefinitionAndOpenModal(formDefinition: any): void {
     this.taskProcessLinkType$.next('form');
     this.formDefinition$.next(formDefinition);
-    this.modal.show();
+    this.modal.open = true;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]: Same as above

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 14]
```diff
@@ -282,8 +289,59 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
     this._subscriptions.add(
       formViewModelComponent.instance.formSubmit.subscribe(() => {
         this.completeTask();
-        this.modal.hide();
+        this.modal.open = false;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]: Same as above

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 15]
```diff
@@ -282,8 +289,59 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
...
+        .getIntermediateSubmission(this.taskInstanceId$.getValue())
+        .subscribe({
+          next: (intermediateSubmission: IntermediateSubmission) => {
+            this.submission$.next({data: intermediateSubmission.submission});
+          },
+          error: () => {
+            this.submission$.next({data: {}});
+          },
+        })
+    );
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]:
```suggestion
      this.taskInstanceId$.pipe(
      switchMap((taskInstanceId: string) => 
      this.taskIntermediateSaveService
        .getIntermediateSubmission(taskInstanceId)
      )
        .subscribe({
          next: (intermediateSubmission: IntermediateSubmission) => {
            this.submission$.next({data: intermediateSubmission.submission});
          },
          error: () => {
            this.submission$.next({data: {}});
          },
        })
    );
```

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 16]
```diff
@@ -282,8 +289,59 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
...
+          error: intermediateSubmission => {
+            this.toastr.error(
+              this.translateService.instant('formManagement.intermediateSave.error')
+            );
+          },
+        })
+    );
+  }
+
+  protected clearCurrentProgress() {
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]:
```suggestion
  protected clearCurrentProgress(): void {
```

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 17]
```diff
@@ -282,8 +289,59 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
...
+            this.submission$.next({data: intermediateSubmission.submission});
+          },
+          error: () => {
+            this.submission$.next({data: {}});
+          },
+        })
+    );
+  }
+
+  protected saveCurrentProgress() {
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]:
```suggestion
  protected saveCurrentProgress(): void {
```

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts [hunk 18]
```diff
@@ -282,8 +289,59 @@ export class TaskDetailModalComponent implements AfterViewInit, OnDestroy {
     this._subscriptions.add(
       formViewModelComponent.instance.formSubmit.subscribe(() => {
         this.completeTask();
-        this.modal.hide();
+        this.modal.open = false;
       })
     );
   }
+
+  private getCurrentProgress() {
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.ts]:
```suggestion
  private getCurrentProgress(): void {
```

File: projects/valtimo/task/src/lib/models/task-intermediate-save.model.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" basis,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export interface IntermediateSubmission {
+  submission: object;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/models/task-intermediate-save.model.ts:18]: Can this be mapped to an interface or is it too generic?

File: projects/valtimo/task/src/lib/models/task-intermediate-save.model.ts [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+  taskInstanceId: string;
+  createdBy: string;
+  createdOn: string;
+  editedBy: string;
+  editedOn: string;
+}
+
+export interface IntermediateSaveRequest {
+  submission: object;
```
Comment (teodor-ritense) [inline projects/valtimo/task/src/lib/models/task-intermediate-save.model.ts:27]: Same question as above

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html (hunk 3 above)
Comment (laurens-ritense) [reply projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.html:55]: No can't do that because then createComponent won't work anymore for form-view-model

File: projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.scss (hunk 5 above)
Comment (laurens-ritense) [reply projects/valtimo/task/src/lib/components/task-detail-modal/task-detail-modal.component.scss:30]: So that the title doesn't wrap too soon

**Review by teodor-ritense — APPROVED** [review]
